### ISSUE TITLE
Covariance matrix is not properly set on Datasets

### ISSUE THREAD

**Issue by registerrier** [gammapy/gammapy#4149]
**Gammapy version**
from at least v0.17 to 1.0rc1

**Bug description**
The covariance matrix is not properly set on `Datasets` when `Models` contain more than one model. Only the sub-covariances of individual models are properly set.  No positive covariance terms between independent model components are set. Thanks @AtreyeeS  for noting this.

**Expected behavior**
Correlations between independent model parameters should be non-zero.

**To Reproduce**
```
import numpy as np
from gammapy.datasets import Datasets
from gammapy.modeling import Covariance

ds = Datasets.read("$GAMMAPY_DATA/fermi-3fhl-crab/Fermi-LAT-3FHL_datasets.yaml",
              filename_models= "$GAMMAPY_DATA/fermi-3fhl-crab/Fermi-LAT-3FHL_models.yaml")

factor_matrix = np.ones((9,9))
ds.models.covariance = Covariance.from_factor_matrix(
        parameters=ds.parameters, matrix=factor_matrix
    )
ds.models.covariance.plot_correlation()
```
yields:
![download-1](https://user-images.githubusercontent.com/5047091/197058054-5829bd1e-c00c-45cf-9018-18de758d4cb2.png)

while:
```
ds[0].models.covariance = Covariance.from_factor_matrix(
        parameters=ds[0].models.parameters, matrix=factor_matrix
    )
ds[0].models.covariance.plot_correlation()
```
yields:
![download-2](https://user-images.githubusercontent.com/5047091/197058469-ebec0ccd-ca05-47c6-afb7-be4efcc3647a.png)

But `ds.models.covariance.plot_correlation()` still plots the incorrect matrix.

**Comment by adonath** [comment]
Thanks @registerrier and @AtreyeeS! I think this is actually expected behavior and there is not really a solution to this until we completely detach the models from the datasets. Look at the following code:

```python3
import numpy as np
from gammapy.datasets import Datasets
from gammapy.modeling import Covariance
from gammapy.modeling.models import Models

ds = Datasets.read("$GAMMAPY_DATA/fermi-3fhl-crab/Fermi-LAT-3FHL_datasets.yaml")

models = Models.read("$GAMMAPY_DATA/fermi-3fhl-crab/Fermi-LAT-3FHL_models.yaml")

factor_matrix = np.ones((9,9))
models.covariance = Covariance.from_factor_matrix(
        parameters=models.parameters, matrix=factor_matrix
)

ds.models = models

models.covariance.plot_correlation()

ds.models.covariance.plot_correlation()

print(models is ds.models)
```

The point here is that `models` and `ds.models` are **not** the same object. `models` is a `Models` object, while `ds.models` is a `DatasetModels` object, which is immutable. What is happening is:
- On `ds.models = models` the individual model components are distributed to the individual datasets (up to bottom)
- On calling `ds.model` again the model components are collected from the individual datasets (bottom to up) and the information on the cross correlation is lost.

**Comment by adonath** [comment]
A possible fix is maybe to cache the `DatasetModels` object in `Datasets`...

**Comment by AtreyeeS** [comment]
@adonath Indeed, `models.covariance.plot_correlation()` gives correct results.
This can be addressed by better documentation maybe?

But I think the issue remains when we have multiple datasets with multiple models. The cross correlation between models of different datasets is not showing correctly

**Comment by AtreyeeS** [comment]
Also, the serialisation is incorrect as the covariance of `Datasets.models` is written

**Comment by adonath** [comment]
@AtreyeeS As I said in the current scheme there is no good solution to this problem. Even caching `DatasetModels` on the `Datasets` objects is problematic, because we need a complex mechanism to detect if the `Datasets` object has been updated, e.g. by `.append(dataset)`. If the appended dataset brings a new model component it has to be registered. The same goes for deletion or setting of datasets... 

I fear we have to live with this limitation for v1.0. But I agree we should probably make a warning / note in the docs.

**Comment by AtreyeeS** [comment]
Shipping a long term version with such limitations is a bit disturbing, no?

(1)
> It is not incorrect, but it is incomplete.

It gives the idea that certain parameters are uncorrelated, so I would say it is erroneous. 

(2)
I think one of the main points of doing a joint fit is to account for correlations between different datasets. And this is being lost completely. For example, in a hypothetical case

```
dataset1 = MapDataset()....
dataset2 = dataset1.copy()

models = Models([sky_model, bkg_dataset1, bkg_dataset2]
datasets = Datasets([dataset1, dataset2])
datasets.models = models
```

The normalisation of the 2 background models are obviously inter-related, but this is not reflected in the fit! This can be quite misleading, I guess.

(3) Also, I am not sure that the behaviour of `Fit.stat_contour` or `Fit.stat_profile` is correct. These rely on `datasets.paramters[par_name]`, and for models containing same parameter names, the behaviour is unpredictable. I think we should rely on `unique_parameters` here

> In general this cannot be fixed without changing the API. The options are:

> Decouple models from Dataset
Then our basic definition of a dataset as something which bundles together counts, irfs, models and the likelihood changes. 

> Make Models and Dataset immutable again, such that .append() etc. does not work anymore
This is more reasonable. I guess if we provide conveniences like `models2 = models1.add_model(new_model)`, etc it would be acceptable for users

**Comment by registerrier** [comment]
This must not be the only issue here.
For instance `DatasetModels(models).covariance` will only contain the covariance from the individual models. The non-diagonal terms are already dropped on init.

All in all the `DatasetModels.covariance` is just the stacking of its individual models covariance. This cannot be correct. No?

**Comment by registerrier** [comment]
Of course, one difficulty is that the covariance matrix is the output of `Fit` and could naturally live only as a `CovarianceResult` object for instance. 
I can't really figure why splitting `Datasets` and `Model` would really solve the issue. In practice, the covariance is really what binds models and a dataset.

Is it really complex to add a mechanism to reset a `Datasets._models` on each mutation of the `Datasets`?

**Comment by adonath** [comment]
Here is a short version of my conclusion from large design discussions we had about this ~1-2 years ago:

> All in all the DatasetModels.covariance is just the stacking of its individual models covariance. This cannot be correct. No?

> And why is the Datasets.models property recreating the DatasetModels on the fly?

The answer is the same here: because `Datasets` are mutable, so can be modified in place using `.extend()` and `.append()` etc. And individual datasets can come with their individual model definitions, which requires an update to the global model. At this point one would need a mechanism to "detect" changes and compare against the cached model object. This is not impossible to implement, but I'm sure we would rather avoid it. It will come with many smaller problems to solve. 

Once we only had a single dataset, we worked bottom->up, which worked fine. However now that we have multiple datasets, this approach does not really work anymore, especially if datasets reside in different processes. Here is one example:

```python3
datasets = Datasets([d1, d2])
d1.models.append(d1_specific_model_component)
d1.models.append(global_model_component) # this component must be visible to d2 as well
```
To keep the bottom->up approach one would need now a mechanism to detect the change in each individual dataset and re-distribute to the other datasets. One could share a single global model state, such that `datasets[idx].models` is always the exact same object. But then one also need a mechanism to handle and update the caching of evaluators in `MapDataset`. 

Or related is the following problem:
```python3
datasets = Datasets([])

new_dataset = MapDataset(..., name="new-dataset")
new_dataset.models = new_specific_model

datasets.append(datasets)

print(datasets.models)

fit = Fit()
fit.run(datasets)
```
This should result in an updated `datasets.models` object. So model changes have to be "monitored" in to directions : bottom-> up and up->bottom. That's a bad design choice. It should be only one of those. 

> I can't really figure why splitting Datasets and Model would really solve the issue. In practice, the covariance is really what binds models and a dataset.

Because once it is fully split, you change completely to a up->bottom approach. There will only be a **single model object** ever, which holds the global state. 

The cleaner way (and we have worked in the direction already) is:

```python3
models = Models()
datasets = Datasets()

new_specific_model = SkyModel(..., dataset_name=["new-dataset"])
models.append(new_specific_model)

new_dataset = MapDataset(..., name="new-dataset")
datasets.append(new_dataset)

fit = Fit()
fit.run(datasets, models)
```
And `datasets[idx].models` are immutable to you do not have to track changes bottom-up. However for parameters and the covariance we still do, because immutability of the parameter values and covariance is either very hard to achieve or more than inconvenient for the users. That's basically how we arrived at the current compromise. However we should not stop there and really work towards splitting the datasets and models to fully implement a up->bottom solution. Models should not be exposed on individual datasets, but that's the main API we have and what people got used to, because we started with single dataset fits only.

**Comment by adonath** [comment]
The reasoning should be here https://docs.gammapy.org/0.20.1/development/pigs/pig-020.html

**Comment by adonath** [comment]
From the dev meeting today. The solution is to add a docs example that shows the following:
```python3

models = Models()
datasets.models = models

fit = Fit()
result = fit.run(datasets)

models.covariance = Covariance(data=result.covariance_result.matrix, parameters=models.parameters)

```
Or similar...

**Comment by adonath** [comment]
Who works on this @registerrier and @AtreyeeS ?

**Comment by registerrier** [comment]
Now that #4152 is merged, the `FitResult.models` copy of the fitted `Models` object contains the actual covariance. How to access it is now shown on the fitting API tutorial.

What is there left to do?
1. Should we expose more the use of `FitResult.models` as the place containing the relevant information? There are several tutorials looking at the `models.to_parameters_table()`, maybe we could use instead the `FitResult.models.to_parameters_table()`? We could also expose writing the `FitResult.models` to disk. 
2. Add an how to, explaining where to find the full covariance matrix?
3. Other?

Opinions @AtreyeeS @adonath ?

**Comment by AtreyeeS** [comment]
It is not very well explained that `FitResult` contains the useful stuff (it was not clear to me till recently!).
Probably exposing `FitResult.models.to_parameters_table()` is a very good idea in the long run.
For now, I would go with adding a HowTo which links to the fitting notebook

**Comment by adonath** [comment]
Fixed by #4152.
